stumpy/mmotifs.py

```diff
@@ -0,0 +1,177 @@
+        Otherwise, this function gets re-routed to its complementary non-normalized
+        equivalent set in the `@core.non_normalized` function decorator.
+
+    p : float, default 2.0
+        The p-norm to apply for computing the Minkowski distance. This parameter is
+        ignored when `normalize == False`.
+
+    Returns
+    -------
+    motif_distances: numpy.ndarray
+        The distances corresponding to a set of subsequence matches for each motif.
+
+    motif_indices: numpy.ndarray
+        The indices corresponding to a set of subsequences matches for each motif.
+
+    motif_subspaces: list
+        A list consisting of arrays that contain the `k`-dimensional
+        subspace for each motif.
+
+    motif_mdls: list
+        A list consisting of arrays that contain the mdl results for
+        finding the dimension of each motif
+    """
+    T = core._preprocess(T)
+    m = T.shape[-1] - P.shape[-1] + 1
+
+    if max_motifs < 1:  # pragma: no cover
+        logger.warning(
+            "The maximum number of motifs, `max_motifs`, "
+            "must be greater than or equal to 1"
+        )
+        logger.warning("`max_motifs` has been set to `1`")
+        max_motifs = 1
+
+    T, M_T, Σ_T = core.preprocess(T, m)
+
+    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
+
+    if max_matches is None:
+        max_matches = np.inf
+
+    if cutoffs is None:
+        P_copy = P.copy().astype(np.float64)
+        P_copy[np.isinf(P_copy)] = np.nan
+        cutoffs = np.nanmax(
+            [
+                np.nanmean(P_copy, axis=1) - 2.0 * np.nanstd(P_copy, axis=1),
+                np.nanmin(P_copy, axis=1),
+            ],
+            axis=0,
+        )
+    if np.isscalar(cutoffs):
+        cutoffs = np.full(P.shape[0], cutoffs)
+
+    motif_distances = []
+    motif_indices = []
+    motif_subspaces = []
+    motif_mdls = []
+
+    candidate_idx = np.argmin(P, axis=1)
+    nn_idx = I[np.arange(len(candidate_idx)), candidate_idx]
+
+    while len(motif_distances) < max_motifs:
+        mdls, subspaces = mdl(T, m, candidate_idx, nn_idx)
+        k = np.argmin(mdls)
+        motif_mdls.append(mdls)
+
+        motif_idx = candidate_idx[k]
+        motif_value = P[k, motif_idx]
+
+        if (
+            motif_value > cutoffs[k]
+            or not np.isfinite(motif_value)
+            or (isinstance(max_distance, float) and motif_value > max_distance)
+        ):
+            break
+
+        motif_subspaces.append(subspaces[k])
+
+        query_matches = match(
+            Q=T[subspaces[k], motif_idx : motif_idx + m],
+            T=T[subspaces[k]],
+            M_T=M_T[subspaces[k]],
+            Σ_T=Σ_T[subspaces[k]],
+            max_matches=max_matches,
+            max_distance=max_distance,
+            atol=atol,
+            # normalize=normalize,
+            # p=p
+        )
+
+        if len(query_matches) > min_neighbors:

```

Comment:
Actually, I think

```
motif_mdls.append(mdls)
motif_subspaces.append(subspaces[k])
```

should be moved here! The length in each should be the same and the way that we have it right now makes it possible for there to be a mismatch in the length, right?

Reply:
Yes, you are right! I had even thought of this with `motif_subspaces.append(subspaces[k])`. Don`t know where I was with the thoughts then when I wrote `motif_mdls.append(mdls)` above the `if` condition.

But does it have to be below `query_matches = match()`? I have now positioned it between the `if` condition and `query_matches = match()`.

Reply:
I think the `append` must all happen at the same time (i.e., within the `if len(query_matches) > min_neighbors:`. This makes the append "all-or-nothing". Having all 4 `append` inside the `if` statement will ensure that they all have the same length. Is that what you are asking?

Reply:
Thank you!